functionary/vllm_monkey_patch/async_llm_engine.py

```diff
@@ -0,0 +1,605 @@
+        """Abort a request during next background loop iteration."""
+        if verbose:
+            logger.info(f"Aborted request {request_id}.")
+
+        self._finished_requests.put_nowait(request_id)
+
+        if (
+            request_id not in self._request_streams
+            or self._request_streams[request_id].finished
+        ):
+            # The request has already finished or been aborted.
+            return
+
+        self._request_streams[request_id].finish()
+
+    def get_new_and_finished_requests(self) -> Tuple[List[dict], Set[str]]:
+        """Get the new requests and finished requests to be
+        sent to the engine."""
+        new_requests: List[dict] = []
+        finished_requests: Set[str] = set()
+
+        while not self._finished_requests.empty():
+            request_id = self._finished_requests.get_nowait()
+            finished_requests.add(request_id)
+            self._request_streams.pop(request_id, None)
+
+        while not self._new_requests.empty():
+            stream, new_request = self._new_requests.get_nowait()
+            if stream.request_id in finished_requests:
+                # The request has already been aborted.
+                stream.finish()
+                continue
+            self._request_streams[stream.request_id] = stream
+            new_requests.append(new_request)
+
+        self.new_requests_event.clear()
+
+        return new_requests, finished_requests
+
+    async def wait_for_new_requests(self):
+        await self.new_requests_event.wait()
+
+
+class _AsyncLLMEngine(LLMEngine):
+    """Extension of LLMEngine to add async methods."""
+
+    # This is a dict mapping request_id to the list of tools/functions details
+    tools_or_functions: dict = {}
+    # This is a dict mapping request_id to the prompt_template_cls
+    prompt_templates: dict = {}
+    # This is a dict mappingg request_id to the generation_state. It contains
+    # the following information:
+    # - stage: one of the following:
+    # ["pre-function", "function", "pre-parameter", "parameter-name", "parameter-value"]
+    # - curr_tokens: all the tokens for the current function call being generated
+    # - curr_text: curr_tokens but in string text form
+    # - func_name: the function name, if any
+    # - param_names: the parameters names, if any
+    gen_states: dict = {}
+
+    def initialize_gen_state(self, request_id):
+        """This function initializes a new generation state if the request_id
+        is not found in self.gen_states
+        """
+        # Prompt template v2 directly begins with "function" stage while
+        # Prompt template v1 needs to generate <|START_OF_FUNCTION_CALL|> first
+        # Future versions are assumed to begin directly with "function" stage
+        self.gen_states[request_id] = {
+            "stage": "function"
+            if self.prompt_templates[request_id].version not in ["v1"]
```

Comment:
Oh I think at first, it is always pre-function?

Reply:
By the way, I think should initialize by an empty dict({}), by this vllm doesn't need to know the structure of state
we will check and initialize inside def grammar_sample
if len(gen_state) == 0:
    gen_state = {...}

Reply:
The start of function token `<|recipient|>` is alr provided in v2 prompt so we basically start from "function" stage instead. I will move this initialization into prompt_template too.

Reply:
Ok